### ISSUE TITLE
Feature request: extending component binding types

### ISSUE THREAD

**Issue by afitiskin** [angular/angular.js#15400]
Currently there are few types of binding data to component:
* `=` – two-way data-binding
* `<` – one-way data-binding
* `@` – one-way data-binding for string 
* `&` – binding for callback-function

So, it's quite enough for basic usage. But when you start use data structures like Immutable.js ability to add your own binding type can make your life much easier.

For example we have some page in your application:
```html
<team-page></team-page>
```

```javascript
// team-page.js
import {List} from 'immutable';

angular.module('team').component('teamPage', {
  templateUrl: '<team-list data="$ctrl.team"></team-list>',
  controller: function () {
    this.team = List({
      name: 'John Smith',
    }, {
      name: 'Jane Norman',
    }, {
      name: 'Andy Cooper',
    });
  }
});
```

You have component `team-list` to render list of team members:
```javascript
// team-list.js
angular.module('team').component('teamList', {
  template: '<ul><li ng-repeat="person in $ctrl.data">{person.name}</li>',
  controller: TeamListController,
  bindings: {
    data: '<',
  },
});

class TeamListController {
  // you have to prepare data for ng-repeat to make it work
  $onChanges() {
    this.data = this.data.toArray(); // use immutable toArray method to convert data to plain Array
  }
}
```

When your whole project uses immutable data structures and one-way data flow (flux) you have only few variants to deal wth it:
* Have a lot of copy-paste code in every component's controller to convert immutable data structures to plain `Arrays` / `Objects`
* Use some kind of directives like [angular-immutable](https://www.npmjs.com/package/angular-immutable), [immutable-angular](immutable-angular), etc. All of these modules uses `$scope` and `$watch` (we prefer not to use it) and have a lot of lacks.
* Use decorators, to decorate components' controllers to parse data from immutable to plain objects

Last variant (using decorators) is interesting and we currently using it, it looks like this:
```javascript
@immutableBindings
class ComponentController {
  static bindings = {
    team: 'I', // symbol I is used for immutable data structure
  };
}

angular.module('some').component('component', {
  template: '<ul><li ng-repeat="person in $ctrl.team"></li>',
  controller: ComponentController,
  bindings: ComponentController.bindings,
})
```

Inside `immutableBindings` decorator we parse `bindings` objects of component's controller, find all `I` (means Immutable) bindings and add necessary conversions to `$onChanges` method. It looks good and works good as well.

Decorators are good, but usage of such kind of decorators looks like a tweak. May be there is a way to extend basic components' binding types more in angular-way? Or may be it is possible to add ability to create custom binding types? What do you think?

**Comment by thorn0** [comment]
Decorators are a great solution. :+1: That's what they are for.

On the other hand, it sounds like `ng-repeat` could be more friendly towards the immutables...

**Comment by afitiskin** [comment]
Few ideas, how to make component `bindings` more friendly.

1. Function binding. 
  ```javascript
  bindings: {
    team: team => team.toArray(),
  }
  ```
  Make one-way data-binding (like `<` does), compare changes on the original `team` object (immutable), but pass to component's controller result of binding-function (plain array).

2. Binding parsers and formatters
  ```javascript
  bindings: {
    team: '<',
    other: '=',
  },
  bindingParsers: {
    team: team => team.toArray(),
    other: data => parseData(data),
  },
  bindingFormatters: {
    other: data => formatData(data),
  }
  ```
  Can be used with different binding-types (one- and two-way). Parsers are helpers to prepare data for component's controller. Formatters are helpers to format data passed from component's controller back to upper scope.

3. Custom bindings
  Create custom binding type:
  ```javascript
  angular.binding('I', ...);
  ```
  Then use it as standard binding type:
  ```javascript
  bindings: {
    team: 'I',
    other: '=',
  }
  ```

**Comment by gkalpak** [comment]
This is an interesting feature with great potential (but not without cost).

First of all, there is some confusion going on. This is not relate to binding types (`=`, `<`, `@` etc). Binding types deal with a different concept (how to pass data between a directive and the outside world).

This feature is about transforming this data when it comes in/before it goes out. It is basically syntactic sugar to help with DRYness and reducing boilerplate (which is a noble goal on its own :smiley:).

--
Regarding the above suggestions:

I would say (1) is out of question.

(3) is interesting (due to its DRYness and declarativeness), but like I said above it mixes two orthogonal concepts: binding type and binding transformation. If we had built-in support for a few common transformations, I could imagine an extended binding DSL that would specify the type of binding, the binding transform and the attribute name across the lines of `<prop-name>: '<type>|<transform>|<attr-name>'` (e.g. `team: '<|I|team'`, `I` standing for `Immutable` tranform). But I don't think we are anywhere near something like that (and it obviously has its cons).

(2) seems powerful and promising, though. We could allow specifying the name of a transformer (e.g. a filter maybe) to be looked up via DI. E.g.:

```js
.filter('toArray', () => {
  return input => input.toArray();
})
.component('myComponent', {
  template: '...',
  bindings: {
    team: '<'
  },
  bindingTransforms: {
    team: 'toArray'
  }
})
```

Another variation would be allowing a binding definition object (but that would be a little verbose):

```js
bindings: {
  team: {
    type: '<',
    transform: 'toArray'
  }
}
```

--
All that said, there will be drawbacks. The first two that come to mind:

1. Increased complexity, both from a maintenance point of view and for users. The DDO is notoriously complex already :smiley:
2. I am not sure how something similar would be achieved in ng2. This is not a blocker, but as a general guideline we avoid giving our users powerful tools that will eventually make a potential transition to ng2 more difficult (unless there is a very good reason).

--
**tl;dr**
Interesting feature, but we need to carefully weight the tradeoffs. If anyone has suggestions/thoughts, please share.

**Comment by kristofdegrave** [comment]
I like the idea and it would definitely reduce the code base, since the $onChanges is used mostly to transform the passed data or provide defaults. But wouldn't this just result in replacing the current $onChanges by this new functionality? I would prefer a well defined lifecycle in this case.

If you would choose to implement this, I would prefer the bindings definition object over a bindingsTranforms object. Advantage of the definition object is that all information of the binding is kept together and you see immediatly how the binding will work.

**Comment by nathraQ** [comment]
First I like the idea of this feature. I would prefer version 1 or 2 just because of how I would think this kind of feature could be implemented in angular2.
I think (didn't try it) it would be possible to write inputDecorators that inherit from the InputDecorator and could change the outcome

**Comment by Anber** [comment]
As Alex said, we use the [decorator](https://gist.github.com/Anber/5f4051fe0958fc07b2533bfa9bec6daa). It's flexible and doesn't request any changes in angular, but it has at least one problem – $onChanges receives an original value. Maybe we can just change implementation of $onChanges and use decorators like this for converting bound data?
